### PR TITLE
Add code enhancement for test module: stig

### PR DESCRIPTION
**Pull request by ShawnHa0**
Modify codes for enhancement of test module stig on Tumbleweed and sle, add comparison of evaluation and remediation baseline, generate stats and logs of baselines changed during remediation for downloading and extended research

- Related ticket: https://progress.opensuse.org/issues/104944
- Needles: N/A
- Verification run: 
    - o3: https://openqa.opensuse.org/tests/2440428#
    - osd: https://openqa.suse.de/tests/9047876#

### CODE REVIEW

**Review by rfan1 — APPROVED** [review]
LGTM

File: data/stig/baseline_comparison.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
+#!/usr/bin/python3
+
+# Extract baseline and save baseline in a hash map
+def extractBaseline(file):
+    stdout = open(file, "r+")
+    line = stdout.readline()
+    tmp = {}
```
Comment (wangqianzou) [inline data/stig/baseline_comparison.py:7]: It is good to use parameter ret instead of tmp, I think.

File: lib/stigtest.pm [hunk 2]
```diff
@@ -44,14 +47,20 @@ sub set_ds_file {
     # Set the ds file for separate product, e.g.,
     # for SLE15 the ds file is "ssg-sle15-ds.xml";
     # for SLE12 the ds file is "ssg-sle12-ds.xml"
+    # for Tumbleweed the ds file is "ssg-opensuse-ds.xml"
```
Comment (wangqianzou) [inline lib/stigtest.pm:50]: It is good to keep the same style by the end of comments( 'comma', 'semicolons','nothing').

File: tests/security/stig/baseline_comparison.pm [hunk 3]
```diff
@@ -0,0 +1,51 @@
+# Copyright 2022 SUSE LLC
+# SPDX-License-Identifier: GPL-2.0-or-Later
+#
+# Summary: Compare the baseline before and after remediation, generate a log
+# Maintainer: shawnhao <weixuan.hao@suse.com>
```
Comment (wangqianzou) [inline tests/security/stig/baseline_comparison.pm:5]: I think you forgot to put poo here  #[104944]

Comment (ShawnHa0) [reply tests/security/stig/baseline_comparison.pm:5]: poo tag added

File: lib/stigtest.pm (hunk 2 above)
Comment (ShawnHa0) [reply lib/stigtest.pm:50]: semicolon added

File: data/stig/baseline_comparison.py (hunk 1 above)
Comment (ShawnHa0) [reply data/stig/baseline_comparison.py]: variable name changed

File: lib/stigtest.pm [hunk 4]
```diff
@@ -32,9 +33,11 @@ our $f_report = 'report.html';
 
 # Set default value for 'scap-security-guide' ds file
 our $f_ssg_sle_ds = '/usr/share/xml/scap/ssg/content/ssg-sle12-ds.xml';
```
Comment (jouyingbin) [inline lib/stigtest.pm:35]: @ShawnHa0 just curious here, will we have ssg-`sle15`-ds.xml?

Comment (ShawnHa0) [reply lib/stigtest.pm:35]: I believe we already have xml for sle15 in oscap-security-guide after installing this package. The specific xml version is determined in the function set_ds_file() below

**Comment by jouyingbin** [comment]
General looks to me, 
would this PR need to be reviewed by @lilyeyes too ?

**Review by jouyingbin — APPROVED** [review]

**Comment by ShawnHa0** [comment]
> General looks to me, would this PR need to be reviewed by @lilyeyes too ?

It would be better to have Lili review this pr but she's on vacation today. I guess it is fine for her to review it next Monday?

File: tests/security/stig/baseline_comparison.pm [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+use Mojo::File 'path';
+
+sub download_baseline_log {
+    # Pass baseline filename & location to this function, download it to local vm and check
+    # Parameter: baseline filename
+    my $filename = $_[0];
+    my $testdata = path('ulogs/' . $filename)->slurp;
+    save_tmp_file($filename, $testdata);
+    assert_script_run("curl -O " . autoinst_url . "/files/" . $filename);
+    assert_script_run("ls | grep $filename");
```
Comment (Amrysliu) [inline tests/security/stig/baseline_comparison.pm:22]: just comment: you could use 'test -e $filename' to check if the file exists.

File: tests/security/stig/oscap_security_guide_setup.pm [hunk 6]
```diff
@@ -26,9 +27,15 @@ sub run {
...
-    record_info("Info", "\"# oscap info $f_ssg_sle_ds\" returns: $out");
+    if (is_sle) {
+        my $f_ssg_sle_ds = $stigtest::f_ssg_sle_ds;
+        $out = script_output("oscap info $f_ssg_sle_ds");
+        record_info("Info", "\"# oscap info $f_ssg_sle_ds\" returns: $out");
+    } else {
+        my $f_ssg_tw_ds = $stigtest::f_ssg_tw_ds;
+        $out = script_output("oscap info $f_ssg_tw_ds");
+        record_info("Info", "\"# oscap info $f_ssg_tw_ds\" returns: $out");
+    }
```
Comment (Amrysliu) [inline tests/security/stig/oscap_security_guide_setup.pm:38]: I would suggest changing to 
```
   my $f_ssg_ds = is_sle ? $stigtest::f_ssg_sle_ds : $stigtest::f_ssg_tw_ds;
   $out = script_output("oscap info $f_ssg_ds");
   record_info("Info", "\"# oscap info $f_ssg_ds\" returns: $out");
```

**Comment by lilyeyes** [comment]
Furthermore, the case is marked with "pass" to with errno=2: http://openqa.suse.de/tests/9009445#step/oscap_xccdf_eval#1/3

**Comment by lilyeyes** [comment]
For O3 verification runs (though passed) but it seems there os no good enough profile to eval/remediate the O3:
For example (# cat oscap_xccdf_eval#1-stdout):
```
--- Starting Evaluation ---

Title
	Verify Group Who Owns passwd File
Rule
	xccdf_org.ssgproject.content_rule_file_groupowner_etc_passwd
Result
	notapplicable

Title
	Verify User Who Owns passwd File
Rule
	xccdf_org.ssgproject.content_rule_file_owner_etc_passwd
Result
	notapplicable

Title
	Verify Permissions on passwd File
Rule
	xccdf_org.ssgproject.content_rule_file_permissions_etc_passwd
Result
	notapplicable
```

**Comment by ShawnHa0** [comment]
> 

I did not change the original code here. And in record_info it contains 'before remediation some fails are permitted', so I guess it is acceptable to be 'pass' here?

**Comment by ShawnHa0** [comment]
> Furthermore, the case is marked with "pass" to with errno=2: http://openqa.suse.de/tests/9009445#step/oscap_xccdf_eval#1/3

I did not change the original code here. And in record_info it contains 'before remediation some fails are permitted', so I guess it is acceptable to be 'pass' here?

**Comment by ShawnHa0** [comment]
> Why test case is marked as "pass" with error=2 "http://openqa.suse.de/tests/9009445#step/oscap_xccdf_remediate/5"

Now failed in latest pr (still running now)

**Review by Amrysliu — APPROVED** [review]
LGTM

**Comment by lilyeyes** [comment]
> > Furthermore, the case is marked with "pass" to with errno=2: http://openqa.suse.de/tests/9009445#step/oscap_xccdf_eval#1/3
> 
> I did not change the original code here. And in record_info it contains 'before remediation some fails are permitted', so I guess it is acceptable to be 'pass' here?

NO.
It is not acceptable as it returns 2 after remediation.

**Pull request merged**